mozilla/Spoke
# Troika Text node

This is the Spoke PR for creating, editing, copying, deleting and exporting Troika Text nodes, based on the troika-three-text package (https://www.npmjs.com/package/troika-three-text). This package has a wide variety of styling options, plus the advantage of being able to display 3D text consistently in Spoke (pure three.js) and Hubs (A-Frame), making it a superior choice for content management purposes versus A-Frame’s `<a-text>` nodes or Mozilla’s hubs-text.js method. There should also be interesting programmatic update use cases for this node.

Overall, 3D artists should not be responsible for _all_ content management, so moving text management out of Blender and into Spoke should help - just like in traditional web dev, you'd build a CMS so that your graphic designer doesn't have to change out copy every single time.

To test:
1. Switch to `feature/troika-text` branch and `yarn install`
2. Create a new scene and add a Troika Text node, customizing the data as desired
3. Save the scene, reload, and see the Troika Text node has persisted
4. Copy the object to see values persist from source object to the copy
5. Delete an object to see text nodes removed from the scene and disposed of
6. Publish the scene and load with this Hubs PR: https://github.com/mozilla/hubs/pull/4855

![image](https://user-images.githubusercontent.com/11968217/143090125-30e6ecdb-720b-4f52-85e9-960dfde6ce29.png)

KNOWN ISSUE: `troika-three-text` has a "Font URL" field, which accepts links to .ttf, .otf and .woff files, downloads them at runtime and converts them to SDF glyphs. Users might expect to be able to upload those file types directly to Spoke, however currently there are both Spoke-side and server-side file type upload limitations that prevents this. For now you can only use remotely-hosted fonts (and watch out for CORS errors), and they cannot be uploaded to Spoke.
